### PR TITLE
User can register a Company Profile

### LINKED ISSUE #5: User registers a company
An authenticated user registers a company with name, website URL, logo (JPEG, PNG) and an e-mail address for contact. 

Each user can only register one company.

All the fields are mandatory. The e-mail must not be the same as the user's e-mail. All the company fields are mandatory.

### PR DESCRIPTION
**Pull request by rmoreno-w**
- Criação da rota para criar um CompanyProfile.
- Criação da rota para exibir os detalhes de um CompanyProfile.
- Implementação de testes para ambas as rotas.

Resolves #5

image

image

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -15,5 +15,6 @@
   # Defines the root path route ("/")
   root "home#index"
 
+  resources :company_profiles, only: [ :index, :new, :create ]
```
Comment (joaorsalmeida) [inline config/routes.rb:18]: You are using `index` to render the current user Company.  It seems that maybe a singular resource would fit perfectly: https://guides.rubyonrails.org/routing.html#singular-resources

**Review by joaorsalmeida — COMMENTED** [review]
It's looking very nice!

File: config/routes.rb (hunk 1 above)
Comment (rmoreno-w) [reply config/routes.rb]: Great catch Joao, my bad! I started this way because at first I thought '_I cannot link to a view that needs an id if I have no id_', so it was a "quick fix". And ended up not correcting it later :') 

I refactored it, removed the index view and changed to the show view on the commit (dac0ff4). Now its as expected form the conventions. To clarify, now the link on the top navbar links to the form to create a profile if the user has no company profile, and to its own company profile if it has one.

File: app/controllers/company_profiles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+class CompanyProfilesController < ApplicationController
+  def new
```
Comment (Bruno-H-Terto) [inline app/controllers/company_profiles_controller.rb:2]: I think the `redirect_to_root_if_user_already_has_a_company_profile` and `return if performed?` methods should be used in the `before_action` callback.

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
+require 'rails_helper'
+
+describe 'User tries to create a company profile' do
```
Comment (Bruno-H-Terto) [inline spec/requests/company_profiles/user_creates_company_profile_request_spec.rb:3]: You can move `type: :request` to the `describe` block instead of `its` blocks.

File: spec/models/company_profile_spec.rb [hunk 4]
```diff
@@ -1,5 +1,69 @@
...
+        user = create(:user)
+        create(:company_profile, user: user)
+        invalid_company_profile = FactoryBot.build(:company_profile, user: user)
+
+        expect(invalid_company_profile).to be_invalid
+      end
+    end
+  end
+
+  describe '#format' do
```
Comment (Bruno-H-Terto) [inline spec/models/company_profile_spec.rb:32]: I think the website url should also validate the format.

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb
Comment (Bruno-H-Terto) [inline spec/requests/company_profiles/user_creates_company_profile_request_spec.rb:1]: I believe it is recommended to add a success scenario.

File: spec/system/company_profiles/user_creates_a_company_profile_spec.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
+require 'rails_helper'
+
+describe 'Registered user tries to access the page to create a company profile' do
```
Comment (Bruno-H-Terto) [inline spec/system/company_profiles/user_creates_a_company_profile_spec.rb:3]: I think it's better to use `type: :system` in the `describe` block and only use `js: true` in the specific `it` blocks that require JavaScript.

File: spec/system/company_profiles/user_creates_a_company_profile_spec.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  it 'but first, has to be signed in', type: :system do
+    create(:user)
+    visit root_path
+
+    expect(page).not_to have_link 'Perfil da Empresa'
+  end
+
+  it 'and succeeds', type: :system, js: true do
```
Comment (Bruno-H-Terto) [inline spec/system/company_profiles/user_creates_a_company_profile_spec.rb:11]: This test seems a bit weak in its expectations. It would be better to use literal strings instead of attributes.

File: spec/models/company_profile_spec.rb [hunk 7]
```diff
@@ -1,5 +1,69 @@
...
+  describe '#valid' do
+    context '#presence' do
+      it { should validate_presence_of(:name)  }
+      it { should validate_presence_of(:website_url)  }
+      it { should validate_presence_of(:contact_email)  }
+      it { should have_one_attached(:logo)  }
+    end
+
+    context '#comparison_with_user_email' do
+      it 'should not be equal to the user email' do
```
Comment (Bruno-H-Terto) [inline spec/models/company_profile_spec.rb:13]: I believe the `contact email` should be unique, but in the current scenario, the test only ensures it is not the same as the email of the user creating the company. The scenario where the email is the same as another user's (other than the one creating the company) is not covered.

File: spec/models/company_profile_spec.rb
Comment (Bruno-H-Terto) [inline spec/models/company_profile_spec.rb:1]: You can use only `build` instead of `FactoryBot.build`.

File: app/views/company_profiles/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    <%= f.text_field :name, class: "relative w-full h-12 py-4 placeholder-transparent transition-all border-b outline-none focus-visible:outline-none peer border-b-slate-200 text-slate-500 autofill:bg-white #{'invalid:border-b-pink-500 invalid:text-pink-500' if @company_profile.errors[:name].any?} focus:border-b-emerald-500 focus:outline-none invalid:focus:border-b-pink-500 disabled:cursor-not-allowed disabled:bg-slate-50 disabled:text-slate-400" %>
+
+    <%= f.label :name, for: 'company_profile_name', class: "cursor-text peer-focus:cursor-default peer-autofill:-top-2 absolute left-0 -top-2 z-[1] text-xs text-slate-400 transition-all before:absolute before:top-0 before:left-0 before:z-[-1] before:block before:h-full before:w-full before:bg-white before:transition-all peer-placeholder-shown:top-2.5 peer-placeholder-shown:text-sm peer-required:after:text-pink-500 peer-required:after:content-['\00a0*'] peer-invalid:text-pink-500 peer-focus:-top-2 peer-focus:text-xs peer-focus:text-emerald-500 peer-invalid:peer-focus:text-pink-500 peer-disabled:cursor-not-allowed peer-disabled:text-slate-400 peer-disabled:before:bg-transparent" %>
+
+    <% if @company_profile.errors[:name].any? %>
```
Comment (Bruno-H-Terto) [inline app/views/company_profiles/_form.html.erb:7]: I think it's better to create a helper for this case.

In the ApplicationHelper, add:

```ruby
def form_error_message(object, attribute)
  if object.errors[attribute].any?
    content_tag :small, class: 'absolute flex justify-between w-full py-1 text-xs transition text-pink-500' do
      content_tag :span do
        object.errors.full_messages_for(attribute).join(", ")
      end
    end
  end
end
```

In the view, replace it with:

```erb
<%= form_error_message(@company_profile, :name) %>
```

**Review by Bruno-H-Terto — COMMENTED** [review]
These screens are looking great!

File: app/models/user.rb
Comment (Bruno-H-Terto) [inline app/models/user.rb:1]: I believe there should be a validation in case a new user registers with an email that is already in use by a company.

File: app/controllers/company_profiles_controller.rb (hunk 2 above)
Comment (rmoreno-w) [reply app/controllers/company_profiles_controller.rb:4]: Sure Bruno! It does indeed help make the actions clearer. Implemented on the commit (b4d65f6)

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb (hunk 3 above)
Comment (rmoreno-w) [reply spec/requests/company_profiles/user_creates_company_profile_request_spec.rb]: Good point, Bruno. I thought about it but forgot to change it 😔 Refactored it in the commit (b4d65f6)

File: spec/models/company_profile_spec.rb (hunk 4 above)
Comment (rmoreno-w) [reply spec/models/company_profile_spec.rb:51]: Another good observation! Implemented in the commit (6174e5f)

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb
Comment (rmoreno-w) [reply spec/requests/company_profiles/user_creates_company_profile_request_spec.rb:1]: Regarding the 'success scenario' test, I didn’t make any changes in the request tests 🥹 These request tests cover *'cases that wouldn’t be possible through direct access in the UI,'* meaning the cases where:

1. A user is not logged in and tries to either access the form or send a `POST` request to create a profile (since a logged-off user doesn't see the company profile button) and
2. The cases where a user who already has a company profile tries to access the form to create one or send a `POST` request to create a new profile (since the company profile button would take them to their own profile and they would not be able to send this request). 

My idea was that the success cases are covered in the system tests, where the user can successfully access the form and create a company profile.

However, your comment did make me go through my code and realize that there were still some untested edge cases after I made changes that João suggested (changing the `index` route to `show`). I have now implemented tests that cover the cases where  in the commit (76c0b4a):

- A logged-off user tries to access the show route.
- A user who has no company profile tries to access the show route.

File: spec/system/company_profiles/user_creates_a_company_profile_spec.rb (hunk 5 above)
Comment (rmoreno-w) [reply spec/system/company_profiles/user_creates_a_company_profile_spec.rb]: As I mentioned in the comment above, refactored in the commit (b4d65f6)

File: spec/system/company_profiles/user_creates_a_company_profile_spec.rb (hunk 6 above)
Comment (rmoreno-w) [reply spec/system/company_profiles/user_creates_a_company_profile_spec.rb]: I agree, using a string in the expectations improve them, makes them more readable and self-contained! Implemented in the commit (76c0b4a)

File: spec/models/company_profile_spec.rb (hunk 7 above)
Comment (rmoreno-w) [reply spec/models/company_profile_spec.rb]: Here, I had not implemented it because the issue did not state that it should be unique. There could be some case where someone from lets say *CampusCode* is responsible for positions for posting both *CampusCode* and *Rebase* positions... But I agree that it is usually a good idea! Maybe @joaorsalmeida could give us his point on this! Anyways, I implemented the validation in the commit (6174e5f)

File: spec/models/company_profile_spec.rb
Comment (rmoreno-w) [reply spec/models/company_profile_spec.rb:1]: Oopsie! My bad again, I removed them where I was using FactoryBot.create, but did not remove for build. Thanks for the tip :) Removed in the commit (b4d65f6)

File: app/views/company_profiles/_form.html.erb (hunk 8 above)
Comment (rmoreno-w) [reply app/views/company_profiles/_form.html.erb:7]: Great idea!! I did not implement it yet as my pull request is getting pretty big, it has changed 36 files already 🥹 but I did take a note and I may fix it along with some other css changes/refactors in my next PR

**Review by Bruno-H-Terto — COMMENTED** [review]
Looks great! Congratulations!

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+      expect(response).to redirect_to new_session_path
+      expect(response.status).to eq 302
+    end
+
+    it 'because they already have a company profile' do
+      user = create(:user)
+      create(:company_profile, user: user)
+      cookie_jar = ActionDispatch::Request.new(Rails.application.env_config.deep_dup).cookie_jar
+      user.sessions.create!.tap do |session|
```
Comment (matheusfsantana) [inline spec/requests/company_profiles/user_creates_company_profile_request_spec.rb:16]: I think it's interesting to move this login logic to a helper

creating a file in `spec/support/login_as` might be a better approach

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+      } })
+
+      expect(response).to redirect_to new_session_path
+      expect(response.status).to eq 302
+    end
+
+    it 'because they already have one' do
+      user = create(:user)
+      create(:company_profile, user: user)
+      cookie_jar = ActionDispatch::Request.new(Rails.application.env_config.deep_dup).cookie_jar
```
Comment (matheusfsantana) [inline spec/requests/company_profiles/user_creates_company_profile_request_spec.rb:48]: Use login helper here too

File: spec/requests/company_profiles/user_sees_company_profile_request_spec.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    get(company_profile_path(company_profile))
+
+    expect(response).to redirect_to new_session_path
+    expect(response.status).to eq 302
+  end
+
+  it 'and fails because they still dont have one' do
+    user = create(:user)
+    cookie_jar = ActionDispatch::Request.new(Rails.application.env_config.deep_dup).cookie_jar
+    user.sessions.create!.tap do |session|
```
Comment (matheusfsantana) [inline spec/requests/company_profiles/user_sees_company_profile_request_spec.rb:17]: Use login helper here too

**Review by matheusfsantana — CHANGES_REQUESTED** [review]

File: app/controllers/company_profiles_controller.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
+class CompanyProfilesController < ApplicationController
+  before_action :redirect_to_root_if_user_already_has_a_company_profile, only: [ :new, :create ]
+  before_action :redirect_to_new_if_user_has_no_company_profile, only: [ :show ]
```
Comment (matheusfsantana) [inline app/controllers/company_profiles_controller.rb:3]: @joaorsalmeida I was wondering, is company registration required to use the system?

So @rmoreno-w I think is a better aproach move this `before action` to `ApplicationController` and change the logic to something like:

```ruby
 def redirect_to_new_if_user_has_no_company_profile
   # check if user is signed in and if is a `regular` user
    if Current.user && Current.user.regular?
        redirect_to new_company_profile_path unless Current.user.company_profile.present?
    end
  end
  ```

File: spec/requests/company_profiles/user_creates_company_profile_request_spec.rb (hunk 10 above)
Comment (rmoreno-w) [reply spec/requests/company_profiles/user_creates_company_profile_request_spec.rb:48]: Agreed! Same as above, took note

File: spec/requests/company_profiles/user_sees_company_profile_request_spec.rb (hunk 11 above)
Comment (rmoreno-w) [reply spec/requests/company_profiles/user_sees_company_profile_request_spec.rb:17]: Agreed! Once more as above, took notes and will fix it

File: app/controllers/company_profiles_controller.rb (hunk 12 above)
Comment (rmoreno-w) [reply app/controllers/company_profiles_controller.rb:3]: I thought about it but, as the issue said *"Each user can only register one company"*, I tried not to make the assumption that registering a company profile was a **must**. If it is indeed, I could refactor it

Comment (joaorsalmeida) [reply app/controllers/company_profiles_controller.rb:3]: Without a company, the user cannot register a job posting. That's the only rule right now.

File: spec/models/company_profile_spec.rb (hunk 7 above)
Comment (joaorsalmeida) [reply spec/models/company_profile_spec.rb]: We can think about that later.

**Review by joaorsalmeida — APPROVED** [review]

**Review by joaorsalmeida — APPROVED** [review]
